src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,350 @@
+                        current_shape[ind] += padl + padr
+                    current_shape = torch.floor((current_shape - (wavelet_length - 1) - 1) / 2 + 1).to(
+                        dtype=torch.int64
+                    )
+                    self.coefficients_shape.extend([current_shape.clone()] * self.n_wavelet_directions)
+                    self.coefficients_padding.extend([(padl.clone(), padr.clone())] * self.n_wavelet_directions)
+
+                self.coefficients_shape = self.coefficients_shape[::-1]
+                self.coefficients_padding = self.coefficients_padding[::-1]
+                self.coefficients_shape.insert(0, self.coefficients_shape[0])  # shape of a/aa/aaa term
+                self.coefficients_padding.insert(0, self.coefficients_padding[0])  # padding of a/aa/aaa term
+
+    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Calculate wavelet coefficients from (image) data.
+
+        Parameters
+        ----------
+        x
+            (Image) data
+
+        Returns
+        -------
+            Wavelet coefficients stacked along one dimension
+
+        Raises
+        ------
+        ValueError
+            If the dimensions along which wavelets are to be calculated are not unique.
+        """
+        # normalize axes to allow negative indexing in input
+        dim = tuple(d % x.ndim for d in self.dim)
+        if len(dim) != len(set(dim)):
+            raise ValueError(f'Axis must be unique. Normalized axis are {dim}')
+
+        # swapping the last axes and the axes to calculate wavelets of
+        x = torch.moveaxis(x, dim, list(range(-len(self.dim), 0)))
+
+        # the ptwt functions work only for real data, thus we handle complex inputs as an additional channel
+        x_real = torch.view_as_real(x).moveaxis(-1, 0) if x.is_complex() else x
+
+        if len(self.dim) == 1:
+            coeffs1 = wavedec(x_real, self.wavelet_name, level=self.level, mode='zero', axis=-1)
+            coefficients_list = self._format_coeffs_1d(coeffs1)
+        elif len(self.dim) == 2:
+            coeffs2 = wavedec2(x_real, self.wavelet_name, level=self.level, mode='zero', axes=(-2, -1))
+            coefficients_list = self._format_coeffs_2d(coeffs2)
+        elif len(self.dim) == 3:
+            coeffs3 = wavedec3(x_real, self.wavelet_name, level=self.level, mode='zero', axes=(-3, -2, -1))
+            coefficients_list = self._format_coeffs_3d(coeffs3)
+
+        # stack multi-resolution wavelets along single dimension
+        coefficients_stack = self._coeff_to_stacked_tensor(coefficients_list)
+        coefficients_stack = (
+            torch.view_as_complex(coefficients_stack.moveaxis(0, -1).contiguous())
+            if x.is_complex()
+            else coefficients_stack
+        )
+
+        # move stacked coefficients to first wavelet dimension
+        return (torch.moveaxis(coefficients_stack, -1, min(dim)),)
```

Comment:
i think we could return a contigous array if we move this before the view_as_complex move, right?

Reply:
yes, only small disadvantage is that if we do it before view_as_complex it needs to be done differently for complex and non-complex data. but this is fine by me.

Reply:
If this is ok for everyone I will add it to the adjoint